File: Tests/ObjC/VSS004_ModelsTests.m
//
//  VSS004_ModelsTests.m
//  VirgilSDK
//
//  Created by Oleksandr Deundiak on 10/13/16.
//  Copyright © 2016 VirgilSecurity. All rights reserved.
//

#import <Foundation/Foundation.h>
#import <XCTest/XCTest.h>

#import "VSSTestsUtils.h"
#import "VSSCrypto.h"

@interface VSS004_ModelsTests : XCTestCase

@property (nonatomic) VSSTestsUtils * __nonnull utils;
@property (nonatomic) VSSTestsConst * __nonnull consts;
@property (nonatomic) VSSCrypto * __nonnull crypto;

@end

@implementation VSS004_ModelsTests

#pragma mark - Setup

- (void)setUp {
    [super setUp];
    
    self.crypto = [[VSSCrypto alloc] init];
    self.consts = [[VSSTestsConst alloc] init];
    self.utils = [[VSSTestsUtils alloc] initWithCrypto:self.crypto consts:self.consts];
}

- (void)tearDown {
    [super tearDown];
}

#pragma mark - Tests

- (void)test001_CreateCardRequestImportExport {
    VSSCreateCardRequest *request = [self.utils instantiateCreateCardRequest];
    
    NSString *exportedData = [request exportData];
    
    VSSCreateCardRequest *importedRequest = [[VSSCreateCardRequest alloc] initWithData:exportedData];
    
    XCTAssert([self.utils checkCreateCardRequest:request isEqualToCreateCardRequest:importedRequest]);
}

- (void)test002_CreateGlobalCardRequestImportExport {
    VSSCreateEmailCardRequest *request = [self.utils instantiateEmailCreateCardRequestWithValidationToken:@"testToken"];
    
    NSString *exportedData = [request exportData];
    
    VSSCreateEmailCardRequest *importedRequest = [[VSSCreateEmailCardRequest alloc] initWithData:exportedData];
    
    XCTAssert([self.utils checkCreateGlobalCardRequest:request isEqualToCreateGlobalCardRequest:importedRequest]);
}

- (void)test003_RevokeCardRequestImportExport {
    VSSRevokeUserCardRequest *revokeRequest = [VSSRevokeUserCardRequest revokeUserCardRequestWithCardId:@"testId" reason:VSSCardRevocationReasonUnspecified];
    
    NSString *exportedData = [revokeRequest exportData];
    
    VSSRevokeCardRequest *importedRevokeRequest = [[VSSRevokeCardRequest alloc] initWithData:exportedData];
    
    XCTAssert([self.utils checkRevokeCardRequest:revokeRequest isEqualToRevokeCardRequest:importedRevokeRequest]);
}

- (void)test004_RevokeGlobalCardRequestImportExport {
    VSSRevokeEmailCardRequest *revokeRequest = [VSSRevokeEmailCardRequest revokeEmailCardRequestWithCardId:@"testId" validationToken:@"testToken" reason:VSSCardRevocationReasonUnspecified];
    
    NSString *exportedData = [revokeRequest exportData];
    
    VSSRevokeEmailCardRequest *importedRevokeRequest = [[VSSRevokeEmailCardRequest alloc] initWithData:exportedData];
    
    XCTAssert([self.utils checkRevokeGlobalCardRequest:revokeRequest isEqualToRevokeGlobalCardRequest:importedRevokeRequest]);
}

- (void)test005_CardImportExport {
    VSSCard *card = [self.utils instantiateCard];
    
    NSString *cardStr = [card exportData];
    
    VSSCard *importedCard = [[VSSCard alloc] initWithData:cardStr];
    
    XCTAssert([self.utils checkCard:card isEqualToCard:importedCard]);
    
    VSSCardValidator *validator = [[VSSCardValidator alloc] initWithCrypto:self.crypto];
    
    VSSPrivateKey *privateKey = [self.crypto importPrivateKeyFromData:[[NSData alloc] initWithBase64EncodedString:self.consts.applicationPrivateKeyBase64 options:0] withPassword:self.consts.applicationPrivateKeyPassword];
    VSSPublicKey *publicKey = [self.crypto extractPublicKeyFromPrivateKey:privateKey];
    NSData *publicKeyData = [self.crypto exportPublicKey:publicKey];
    
    XCTAssert([validator addVerifierWithId:self.consts.applicationId publicKeyData:publicKeyData]);
    
    XCTAssert([validator validateCardResponse:importedCard.cardResponse]);
}

@end

File: Tests/VSS004_ModelsTests.swift
//
//  VSS004_ModelsTests.swift
//  VirgilSDK
//
//  Created by Oleksandr Deundiak on 10/17/16.
//  Copyright © 2016 VirgilSecurity. All rights reserved.
//

import Foundation
import VirgilSDK
import XCTest

class VSS004_ModelsTests: XCTestCase {
    private var utils: VSSTestUtils!
    private var consts: VSSTestsConst!
    private var crypto: VSSCrypto!
    
    // MARK: Setup
    
    override func setUp() {
        super.setUp()
        
        self.consts = VSSTestsConst()
        self.crypto = VSSCrypto()
        self.utils = VSSTestUtils(crypto: self.crypto, consts: self.consts)
    }
    
    override func tearDown() {
        self.utils = nil
        
        super.tearDown()
    }
    
    func test001_CreateCardRequestImportExport() {
        let request = self.utils.instantiateCreateCardRequest()
        
        let exportedData = request.exportData()
        
        let importedRequest = VSSCreateCardRequest(data: exportedData)!
        
        XCTAssert(self.utils.check(createCardRequest: request, isEqualToCreateCardRequest: importedRequest))
    }
    
    func test002_CreateGlobalCardRequestImportExport() {
        let request = self.utils.instantiateEmailCreateCardRequest(withIdentity: "testIdentity", validationToken: "testToken", keyPair: nil)
        
        let exportedData = request.exportData()
        
        let importedRequest = VSSCreateEmailCardRequest(data: exportedData)!
        
        XCTAssert(self.utils.check(createGlobalCardRequest: request, isEqualToCreateGlobalCardRequest: importedRequest))
    }

    func test003_RevokeCardRequestImportExport() {
        let revokeRequest = VSSRevokeUserCardRequest(cardId: "testId", reason: .unspecified)
        
        let exportedData = revokeRequest.exportData()
        
        let importedRevokeRequest = VSSRevokeUserCardRequest(data: exportedData)!
        
        XCTAssert(self.utils.check(revokeCardRequest: revokeRequest, isEqualToRevokeCardRequest: importedRevokeRequest))
    }
    
    func test004_RevokeGlobalCardRequestImportExport() {
        let revokeRequest = VSSRevokeEmailCardRequest(cardId: "testId", validationToken:"@testToken", reason: .unspecified)
        
        let exportedData = revokeRequest.exportData()
        
        let importedRevokeRequest = VSSRevokeEmailCardRequest(data: exportedData)!
        
        XCTAssert(self.utils.check(revokeGlobalCardRequest: revokeRequest, isEqualToRevokeGlobalCardRequest: importedRevokeRequest))
    }
    
    func test005_CardImportExport() {
        let card = utils.instantiateCard()
        
        let cardStr = card.exportData()
        
        let importedCard = VSSCard(data: cardStr)!
        
        XCTAssert(self.utils.check(card: card, isEqualToCard: importedCard))
    }
}
